src/installer/image/installer.go

```diff
@@ -2,82 +2,131 @@ package image
 }
 
-func NewImageInstaller(fs afero.Fs, props *Properties) *imageInstaller {
-	return &imageInstaller{
+func NewImageInstaller(fs afero.Fs, props *Properties) *ImageInstaller {
+	return &ImageInstaller{
 		fs:    fs,
 		props: props,
 	}
 }
 
-type imageInstaller struct {
+type ImageInstaller struct {
 	fs    afero.Fs
 	props *Properties
 }
 
-func (installer *imageInstaller) InstallAgent(targetDir string) error {
-	log.Info("installing agent", "target dir", targetDir)
-	if err := installer.installAgentFromImage(targetDir); err != nil {
+func (installer ImageInstaller) ImageDigest() string {
+	return installer.props.ImageDigest
+}
+
+func (installer *ImageInstaller) InstallAgent(targetDir string) error {
+	log.Info("installing agent from image")
+	err := installer.fs.MkdirAll(targetDir, 0755)
+	if err != nil {
+		log.Info("failed to create install target dir", "err", err, "targetDir", targetDir)
+		return errors.WithStack(err)
+	}
+	if err := installer.installAgentFromImage(); err != nil {
 		_ = installer.fs.RemoveAll(targetDir)
-		return fmt.Errorf("failed to install agent: %w", err)
+		log.Info("failed to install agent from image", "err", err)
+		return errors.WithStack(err)
 	}
-	return symlink.CreateSymlinkForCurrentVersionIfNotExists(installer.fs, targetDir)
+	sharedDir := installer.props.PathResolver.AgentSharedBinaryDirForDigest(installer.props.ImageDigest)
+	return symlink.CreateSymlinkForCurrentVersionIfNotExists(installer.fs, sharedDir)
 }
 
-func (installer *imageInstaller) UpdateProcessModuleConfig(targetDir string, processModuleConfig *dtypes.ProcessModuleConfig) error {
-	return processmoduleconfig.UpdateProcessModuleConfig(installer.fs, targetDir, processModuleConfig)
+func (installer ImageInstaller) UpdateProcessModuleConfig(targetDir string, processModuleConfig *dtypes.ProcessModuleConfig) error {
+	sourceDir := installer.props.PathResolver.AgentSharedBinaryDirForDigest(installer.props.ImageDigest)
+	return processmoduleconfig.CreateAgentConfigDir(installer.fs, targetDir, sourceDir, processModuleConfig)
 }
 
-func (installer *imageInstaller) installAgentFromImage(targetDir string) error {
-	_ = installer.fs.MkdirAll(CacheDir, 0755)
+func (installer *ImageInstaller) installAgentFromImage() error {
+	defer installer.fs.RemoveAll(CacheDir)
+	err := installer.fs.MkdirAll(CacheDir, 0755)
+	if err != nil {
+		log.Info("failed to create cache dir", "err", err)
+		return errors.WithStack(err)
+	}
 	image := installer.props.ImageUri
 
 	sourceCtx, sourceRef, err := getSourceInfo(CacheDir, *installer.props)
 	if err != nil {
 		log.Info("failed to get source information", "image", image)
-		return err
+		return errors.WithStack(err)
 	}
 
 	imageDigest, err := getImageDigest(sourceCtx, sourceRef)
 	if err != nil {
 		log.Info("failed to get image digest", "image", image)
-		return err
+		return errors.WithStack(err)
+	}
+
+	imageDigestEncoded := imageDigest.Encoded()
+	imageCacheDir := filepath.Join(CacheDir, imageDigestEncoded)
```

Comment:
filepath joins should be wrapped in a function as their level of abstraction is different then the one of `installAgentFromImage`